.teamcity/src/main/kotlin/util/RerunFlakyTest.kt

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package util
+
+import common.BuildToolBuildJvm
+import common.Os
+import common.applyDefaultSettings
+import common.buildToolGradleParameters
+import common.checkCleanM2
+import common.compileAllDependency
+import common.functionalTestExtraParameters
+import common.functionalTestParameters
+import common.gradleWrapper
+import common.killProcessStep
+import configurations.CompileAll
+import jetbrains.buildServer.configs.kotlin.v2019_2.BuildStep
+import jetbrains.buildServer.configs.kotlin.v2019_2.BuildType
+import jetbrains.buildServer.configs.kotlin.v2019_2.ParameterDisplay
+
+class RerunFlakyTest(os: Os) : BuildType({
+    val id = "Util_RerunFlakyTest${os.asName()}"
+    name = "Rerun Flaky Test - ${os.asName()}"
+    description = "Allows you to rerun a selected flaky test 10 times"
+    id(id)
+    val testJvmVendorParameter = "testJavaVendor"
+    val testJvmVersionParameter = "testJavaVersion"
+    val testTaskParameterName = "testTask"
+    val testTaskOptionsParameterName = "testTaskOptions"
+    val daemon = true
+    applyDefaultSettings(os, BuildToolBuildJvm, 30)
+    val extraParameters = functionalTestExtraParameters("RerunFlakyTest", os, "%$testJvmVersionParameter%", "%$testJvmVendorParameter%")
+    val parameters = (
+        buildToolGradleParameters(daemon) +
+            listOf(extraParameters) +
+            functionalTestParameters(os)
+        ).joinToString(separator = " ")
+
+    killProcessStep("KILL_LEAKED_PROCESSES_FROM_PREVIOUS_BUILDS", daemon, os)
+    steps {
+        gradleWrapper {
+            name = "SHOW_TOOLCHAINS"
+            tasks = "javaToolchains"
+            gradleParams = parameters
+        }
+    }
+    (1..10).forEach { idx ->
```

Comment:
How about making `10` configurable?

Reply:
You mean as a parameter for the class `RerunFlakyTest`? At runtime it would be great to have it configurable, though I didn't find a way to dynamically re-run a build step a number of times. Do you have some ideas there?

Reply:
Maybe wrap it into a shell loop?

```
#!/bin/bash

set -x

x=1
while [ $x -le %IterationNumber% ]
do
  ./gradlew test --tests XX --rerun
  x=$(( $x + 1 ))
done
```

Reply:
I somewhat wanted to use the Gradle build step. If a do a shell loop, then I'll lose some of the advantages of that. If you run your build on a branch, you can change the number of retries there, so I don't think this is a big issue. Let's see how folks use it in practice.